contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,396 @@
+            deltaAmounts[i] += deltaIn;
+            deltaAmounts[i + 1] += deltaOut;
+
+            prevAmountOut = zeroForOne ? uint128(-curDeltas.amount1()) : uint128(-curDeltas.amount0());
+            prevCurrencyOut = params.path[i].intermediateCurrency;
+            sqrtPriceX96AfterList[i] = sqrtPriceX96After;
+            initializedTicksLoadedList[i] =
+                PoolTicksCounter.countInitializedTicksLoaded(manager, poolKey, tickBefore, tickAfter);
+        }
+    }
+
+    function _quoteExactInputSingle(ExactInputSingleParams memory params)
+        private
+        returns (BalanceDelta deltas, uint160 sqrtPriceX96After, int24 tickAfter)
+    {
+        return _quoteExact(
+            params.poolKey,
+            params.zeroForOne,
+            int256(int128(params.amountIn)),
+            params.sqrtPriceLimitX96,
+            params.hookData
+        );
+    }
+
+    function _quoteExactOutput(ExactOutputParams memory params)
+        private
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksLoadedList
+        )
+    {
+        uint256 pathLength = params.path.length;
+
+        deltaAmounts = new int128[](pathLength + 1);
+        sqrtPriceX96AfterList = new uint160[](pathLength);
+        initializedTicksLoadedList = new uint32[](pathLength);
+        Currency prevCurrencyIn;
+        uint128 prevAmountIn;
+
+        for (uint256 i = pathLength; i > 0; i--) {
+            (PoolKey memory poolKey, bool oneForZero) = SwapIntention.getPoolAndSwapDirection(
+                params.path[i - 1], i == pathLength ? params.currencyOut : prevCurrencyIn
+            );
+
+            (, int24 tickBefore,,) = manager.getSlot0(poolKey.toId());
+
+            ExactOutputSingleParams memory singleParams = ExactOutputSingleParams({
+                poolKey: poolKey,
+                zeroForOne: !oneForZero,
+                recipient: params.recipient,
+                amountOut: i == pathLength ? params.amountOut : prevAmountIn,
+                sqrtPriceLimitX96: 0,
+                hookData: params.path[i - 1].hookData
+            });
+            (BalanceDelta curDeltas, uint160 sqrtPriceX96After, int24 tickAfter) = _quoteExactOutputSingle(singleParams);
+
+            (int128 deltaIn, int128 deltaOut) =
+                !oneForZero ? (curDeltas.amount0(), curDeltas.amount1()) : (curDeltas.amount1(), curDeltas.amount0());
+            deltaAmounts[i - 1] += deltaIn;
+            deltaAmounts[i] += deltaOut;
+
+            prevAmountIn = !oneForZero ? uint128(curDeltas.amount0()) : uint128(curDeltas.amount1());
+            prevCurrencyIn = params.path[i - 1].intermediateCurrency;
+            sqrtPriceX96AfterList[i - 1] = sqrtPriceX96After;
+            initializedTicksLoadedList[i - 1] =
+                PoolTicksCounter.countInitializedTicksLoaded(manager, poolKey, tickBefore, tickAfter);
```

Comment:
can someone double check this is correct? I'm like 99% sure it is (since we work backwards through the path, but store those values in the same order as what would happen for an actual swap)

Reply:
whats the concern?

Reply:
Resolved offline, it's different behavior from v3 because v3 requires that the path is passed in backwards whereas here we pass it in normally and iterate backwards